### DISCUSSION TITLE
zerovec slower than cow?

### DISCUSSION THREAD

**Discussion by Thomblin** [unicode-org/icu4x#4515] [Q&A] [answered]
I found this project through https://manishearth.github.io/blog/2022/08/03/zero-copy-2-zero-copy-all-the-things/ and was trying some benchmark using `quick_xml::de::from_str` expecting that `zerotovec `is much faster then using `String `or `Cow `regarding the numbers on https://docs.rs/zerovec/latest/zerovec/

I am just wondering: Did I misunderstand something or did I implement something the wrong way?

Basically what I am doing is:

`let r: Root = quick_xml::de::from_str(xml).unwrap();`

where `Root `is a serde struct like

```
#[derive(Serialize, Deserialize)]
pub struct Root{
    #[serde(rename = "$text")]
    pub text: Option<String>,
    #[serde(rename = "@Status")]
    pub status: String,
}
```

for a 5 to 80 MB xml file containing around 60 different nested elements.

For the benchmark I replace all `String `occurences with either
`Cow<'a, str>` or `VarZeroVec<'data, str>`
and adding the `serde(borrow)` directive

However I am suprised by the benchmark:

> test benches::bench_parse_to_cow     ... bench:     368,994 ns/iter (+/- 175,561)
> test benches::bench_parse_to_string  ... bench:     400,821 ns/iter (+/- 174,372)
> test benches::bench_parse_to_zerovec ... bench:     620,131 ns/iter (+/- 208,128)

So my question is: Is this expected, or do I need to change something to gain the full advantage of zerotovec?

**Comment by Manishearth** [comment, marked as answer]
I don't understand, `VarZeroVec<str>` is not the equivalent of `Cow<str>` and `String`. It's the equivalent of `Vec<String>` (which cannot be represented in `Cow` in a way that will not allocate on deserialization).

There's no advantage to using ZeroVec over Cow: ZeroVec supports doing _more things_ in zero-copy. If you can do it in Cow without allocating you don't need this crate.

**Comment by sffc** [comment]
XML supports character escape sequences, so it is not generally possible to get a `&str`, and Serde does not support conditionally parsing as borrowed or owned based on input ([citation](https://stackoverflow.com/questions/63201624/how-do-you-make-a-serde-struct-work-with-both-borrowed-and-owned-data)), so it won't be possible to improve deserialization performance of XML using zerovec.

**Reply by Manishearth** [reply]
Oh, I didn't even notice that.

Yeah zerovec only helps in situations where zero copy deserialization is possible, which typically means binary serialization formats.

**Comment by Thomblin** [comment]
Thanks a lot for your input. Still learning how the details of Rust are working. I guess I need to read some more

@Manishearth I am not sure if I understand your last comment correctly. Does quick_xml::de with serde use  zero copy deserialization if I choose Cow<'a, str> as type?

**Reply by Manishearth** [reply]
I believe it does not.
